xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+      startRpcStream();
+    }
+    adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
+  }
+
+  /**
+   * Builds a channel to the given server URI with the first supported channel creds config.
+   */
+  private static ManagedChannel buildChannel(String serverUri,List<ChannelCreds> channelCredsList) {
+    ManagedChannel ch = null;
+    // Use the first supported channel credentials configuration.
+    // Currently, only "google_default" is supported.
+    for (ChannelCreds creds : channelCredsList) {
+      if (creds.getType().equals("google_default")) {
+        ch = GoogleDefaultChannelBuilder.forTarget(serverUri).build();
+        break;
+      }
+    }
+    if (ch == null) {
+      ch = ManagedChannelBuilder.forTarget(serverUri).build();
+    }
+    return ch;
+  }
+
+  /**
+   * Establishes the RPC connection by creating a new RPC stream on the given channel for
+   * xDS protocol communication.
+   */
+  private void startRpcStream() {
+    checkState(adsStream == null, "Previous adsStream has not been cleared yet");
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
+        AggregatedDiscoveryServiceGrpc.newStub(channel);
+    adsStream = new AdsStream(stub);
+    adsStream.start();
+    stopwatch.reset().start();
+  }
+
+  /**
+   * Terminates the RPC connection.
+   */
+  @VisibleForTesting
+  void shutdownRpcStream() {
+    if (adsStream != null) {
+      adsStream.close(Status.CANCELLED.withDescription("shutdown").asException());
+    }
+    if (rpcRetryTimer != null) {
+      rpcRetryTimer.cancel();
+    }
+  }
+
+  /**
+   * Handles LDS response to find the HttpConnectionManager message for the requested resource name.
+   * Proceed with the resolved RouteConfiguration in HttpConnectionManager message of the requested
+   * listener, if exists, to find the VirtualHost configuration for the "xds:" URI
+   * (with the port, if any, stripped off). Or sends an RDS request if configured for dynamic
+   * resolution. The response is NACKed if contains invalid data for gRPC's usage. Otherwise, an
+   * ACK request is sent to management server.
+   */
+  private void handleLdsResponse(DiscoveryResponse ldsResponse) {
+    logger.log(Level.FINE, "Received a LDS response: {0}", ldsResponse);
+    checkState(ldsResourceName != null && configWatcher != null,
+        "No LDS request was ever sent. Management server is doing something wrong");
+    adsStream.ldsRespNonce = ldsResponse.getNonce();
+
+    // Unpack Listener messages.
+    Map<String, Listener> listeners = new HashMap<>(ldsResponse.getResourcesCount());
+    try {
+      for (com.google.protobuf.Any res : ldsResponse.getResourcesList()) {
+        Listener l = res.unpack(Listener.class);
+        listeners.put(l.getName(), l);
+      }
+    } catch (InvalidProtocolBufferException e) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+          ldsResponse.getNonce(), "Broken LDS response");
```

Comment:
As mentioned for the `sendNackRequest` comment, consider adding a `Code` value that Nack expects in its error detail. The error message can also be more descriptive and relevant than just saying "Broken LDS response". Considering focus on debuggability/logging we should be careful in these aspects.

Reply:
For general error message, as I mentioned in https://github.com/grpc/grpc-java/pull/6320#discussion_r346972032, it is a tradeoff between verbosity and scope narrowed down for locating the issue. We leverage to provide error message that allows TD to somewhat narrow down the debug scope without introduce too much verbosity in our code. 

For this specific one, we cannot tell much other than "Broken LDS response" for getting `InvalidProtocolBufferException`.

Reply:
This particular case `InvalidProtocolBufferException` is almost never likely to happen unless the xDS server has sent you bad data. But I strongly urge you to include a `Code` value. At least look at Envoy to see what Code values it uses (or doesn't)

Reply:
For this particular case, the NACK request will contain error_detail with Code `INVALID_ARGUMENT` and an error message of "Broken LDS response". Actually, all of NACK request we send back to TD will contain Code `INVALID_ARGUMENT`. Given the Code enum in [`com.google.rpc.Code`](https://github.com/googleapis/googleapis/blob/d8dd7fe8d5304f7bd1c52207703d7f27d5328c5a/google/rpc/code.proto#L32), this is the only suitable Code for all of our cases that response is broken or contains invalid data.

Reply:
May be for all the current cases `INVALID_ARGUMENT` applies but I can see the following potentially applying as well (for our XdsClient talking to xDS server): `NOT_FOUND`, `RESOURCE_EXHAUSTED`, `FAILED_PRECONDITION`, `ABORTED`, `OUT_OF_RANGE` etc.

Reply:
If you read the comment for those `Code`s, none of them is really suitable for us. For example, `RESOURCE_EXHAUSTED` is for "Some resource has been exhausted, perhaps a per-user quota, or perhaps the entire file system is out of space", which is completely not true. Actually, none of the `Code`s really works for us, that's why I did not want to set it. `error_detail` is used for TD debugging purposes, `Code` doesn't give TD any debugging benefit. I would say we should be focusing on the error message.

Reply:
Yes, that's why I said they could potentially apply but not specifically in the current XdsClientImpl code. However a few things I can say are:
- since this code appears on the TD side it will be useful for them to log/filter/search etc
- in future we cannot rule out having codes other than `INVALID_ARGUMENT`
- it will be good to do what other clients do (Envoy?)